src/scripts/setup_test_data/setup_test_data.py

```diff
@@ -0,0 +1,118 @@
+'''
+This script creates a Synpase project, connects to an existing S3 bucket,
+and syncs test data files to the project.
+'''
+import argparse
+import logging
+import sys
+
+import boto3
+import synapseclient
+from synapseformation import client as synapseformation_client
+import synapseutils
+
+
+project_name = 'BridgeDownstreamTest'
+
+logging.basicConfig()
+log = logging.getLogger(__name__)
+log.setLevel(logging.DEBUG)
+
+
+def get_synapse_client(ssm_parameter):
+  '''Get an instance of the synapse client'''
+  ssm_client = boto3.client('ssm')
+  token = ssm_client.get_parameter(
+    Name=ssm_parameter,
+    WithDecryption=True)
+  syn = synapseclient.Synapse()
+  syn.login(authToken=token['Parameter']['Value'])
+  return syn
+
+
+def get_project_id(syn, principal_id):
+  '''Get the id of the synapse project if it exists'''
+  log.info(f'Get project id for {project_name}')
+  projects = syn.restGET(f'/projects/user/{principal_id}')
+  BridgeDownstreamTest = next(
+    filter(
+      lambda x: x['name'] == project_name,
+      projects.get('results')
+      ),
+    None)
+  return '' if BridgeDownstreamTest is None else BridgeDownstreamTest.get('id')
+
+
+def create_project(syn, template_path):
+  '''Create a synapse project from a template'''
+  log.info(f'Create project {project_name}')
+  try:
+    response = synapseformation_client.create_synapse_resources(template_path)
+    if response is not None:
+      return response.get('id')
+  except Exception as e:
+      print(e)
+      sys.exit(1)
+
+
+def setup_external_storage(syn, bucket_name, project_id, synapse_folder_name):
+  '''Connect bucket as external storage for the Synapse project'''
+  log.info(f'Use s3 bucket {bucket_name} as storage for {project_name}')
+  storage_location = syn.create_s3_storage_location(
+          parent = project_id,
+          folder_name = synapse_folder_name,
+          bucket_name = bucket_name)
+  storage_location_info = {
+         k: v for k, v in
+         zip(['synapse_folder', 'storage_location', 'synapse_project'],
+             storage_location)}
+  return(storage_location_info)
+
+
+def get_folder_id(syn, project_id, synapse_folder_name):
+  log.info(f'Get synapse id for {synapse_folder_name} folder')
+  response = syn.getChildren(project_id, includeTypes=['folder'])
+  folder = next(
+    filter(
+      lambda x: x['name'] == synapse_folder_name,
+      response
+      ),
+    None)
+  return '' if folder is None else folder.get('id')
+
+def main():
+
+  log.info(f'Begin setting up test data')
+
+  # get synapse client
+  ssm_parameter = 'synapse-bridgedownstream-auth'
+  syn = get_synapse_client(ssm_parameter=ssm_parameter)
+
+  # see if project exists and get its id
+  principal_id = '3432808' # BridgeDownstream Synapse service account
+  project_id = get_project_id(syn, principal_id)
+  if not project_id:
+    template_path = './src/scripts/setup_test_data/synapse-formation.yaml'
+    project_id = create_project(syn, template_path)
```

Comment:
I suspect if you had to re-run this code, it would be fail because create_resource currently doesn't handle updating of resources.  An error is thrown if the project already exists.

Reply:
If I had to make an *update*, yes, that would be a problem. I've been resorting to manually deleting the project in that case. As it is everything runs great b/c I check whether I was able to retrieve a project id and only run create_project if not.

So that leads me to several things I have been thinking I would like from synapseformation:
1. get_or_create instead of just create. I know of "only_get" but I think that requires that I edit the template or something...and I want the client to be smart for me, not to do this in my script
2. a warning for create when the project already exists instead of an error
3. an update method

Reply:
Ahh, yes I see that part of the code now. Thanks for the suggestions.

I think the client currently fills out the template with IDs and writes it back, I can see how that's not ideal as the client should just be smart enough to find the resources on synapse if they already exist (technically synapse can do this)

When you store entities with the same name in synapse, it will automatically get the entity if it already exists.  I should draft up some design docs for how synapseformation should work and have it as something the team can implement.